### ISSUE TITLE
DNA-mapping stalled

### ISSUE THREAD

**Issue by ppnin** [maxplanck-ie/snakepipes#391]
Hi,

Yesterday afternoon I started the `DNA-mapping` procedure and left my computer working over night. Now I see that the mapping is still not finished, and I wonder if it stuck on one of the jobs (no progress since yesterday evening):

```
[Tue Feb 26 20:14:28 2019]
rule bamCoverage_filtered:
    input: filtered_bam/P10858_1001_tr1.filtered.bam, filtered_bam/P10858_1001_tr1.filtered.bam.bai
    output: bamCoverage/P10858_1001_tr1.filtered.seq_depth_norm.bw
    log: bamCoverage/logs/bamCoverage.P10858_1001_tr1.filtered.out, bamCoverage/logs/bamCoverage.P10858_1001_tr1.filtered.err
    jobid: 25
    benchmark: bamCoverage/.benchmark/bamCoverage.P10858_1001_tr1.filtered.benchmark
    wildcards: sample=P10858_1001_tr1
    threads: 10

Activating conda environment: /home/pavle/miniconda3/envs/974433cd7e65c030b6b99bf2199033d1
...
```
Is there a way to check what's going on with it? Despite `threads: 10` there is no activity.
Thanks
/Pavle

**Comment by dpryan79** [comment]
Are you running it locally or via a cluster? Usually when this happens on a cluster the scheduler killed a job for some reason (usually memory related, we just increase the memory requested then).

**Comment by ppnin** [comment]
I run it locally, as `DNA-mapping -i $inp_dir -o $out_dir --local -j 10 hg19`  ( I also ran it few days ago on the same samples (although raw) and there was no problem).

**Comment by dpryan79** [comment]
Hmm, that's odd. Once in a blue moon we've run into bowtie2 becoming a zombie. I would suggest killing that process if you can and just restarting snakePipes.

**Comment by steffenheyne** [comment]
maybe look with `top` first and check if something from the pipeline is still running (%CPU)
otherwise just do what @dpryan79 mentioned

**Comment by ppnin** [comment]
Alright, I think the problem is related to RAM. I restarted snakePipes and `DNA-mapping`, and kept an eye on the machine resources...RAM usage goes up to the top (32GB on this machine). I guess that kills the job, as @dpryan79 said. I just wonder why it didn't happen when i ran it before..   generally, how memory intensive is the `rule bamCoverage_filtered`?

**Comment by steffenheyne** [comment]
hard to say, mainly data specific. while running on the cluster, that rule is currently limited to 16 threads *4GB (per thread) = 64GB and this seems fine for many datasets we have where. 
Running locally, you can try to lower the -j value which should result in lower memory usage overall.

**Comment by dpryan79** [comment]
We really need to track down why deepTools uses so much memory. I'm tempted to rewrite the whole thing in Rust just because of that :(

**Comment by steffenheyne** [comment]
uhhh :-)
memory footprint could be also related to pysam...

**Comment by dpryan79** [comment]
Could be, but the overhead shouldn't be so absurdly high. Anyway, that's a topic for the deepTools repository.

**Comment by ppnin** [comment]
@steffenheyne  that's good idea, I'll test now with `-j 7` :)

**Comment by ppnin** [comment]
It didn't help, unfortunately...  later I even decreased it to `-j 5`, but after a minute (or so)  RAM usage goes to maximum :/

**Comment by dpryan79** [comment]
Yikes, how big is the BAM file at that point?

**Comment by ppnin** [comment]
It's 3.2 GB

**Comment by dpryan79** [comment]
That's a pretty reasonable size. Are you using a particularly large blacklist by chance?

**Comment by ppnin** [comment]
Not that I know of, I don't see it in `DNA-maping` options - am I missing something? The only difference between the samples that I use now and those from few days ago, when DNA-mapping` went without a problem, is that I removed adapters and trimmed first/last few bp (after checking FastQC reports).

**Comment by dpryan79** [comment]
Can you make the most problematic BAM file (privately) available and the exact DNA-mapping parameters you're using? I'd be interested in debugging why this is blowing up the memory so much.

**Comment by ppnin** [comment]
I believe I can (but I have to check it first).  I could send you a link to the file to you e-mail address, if that's fine?

**Comment by ppnin** [comment]
I sent it now :)

**Comment by dpryan79** [comment]
For those wondering, this was caused by deepTools "caching" the reads in a genomic chunk in memory. In this case 70% of the reads happened to be in one chunk, which resulted in 34GB of allocated memory. For deepTools 3.2.1 I've restructured things to drastically decrease memory use, at the cost of a small increase in run time. In general that should make it much friendlier to use with standard resources.
